pkg/object/tcpproxy/connection.go

```diff
@@ -0,0 +1,395 @@
+		})
+	})
+}
+
+func (c *Connection) goWithRecover(handler func(), recoverHandler func(r interface{})) {
+	go func() {
+		defer func() {
+			if r := recover(); r != nil {
+				logger.Errorf("tcp connection goroutine panic: %v\n%s\n", r, string(debug.Stack()))
+				if recoverHandler != nil {
+					// it is not needed to wrap recoverHandler with go func in the current scenario
+					recoverHandler(r)
+				}
+			}
+		}()
+		handler()
+	}()
+}
+
+// Write receive other connection data
+func (c *Connection) Write(buf *iobufferpool.StreamBuffer) (err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			logger.Errorf("tcp connection has closed, local addr: %s, remote addr: %s, err: %+v",
+				c.localAddr.String(), c.remoteAddr.String(), r)
+			err = ErrConnectionHasClosed
+		}
+	}()
+
+	select {
+	case c.writeBufferChan <- buf:
+		return
+	default:
+	}
+
+	// try to send data again in 60 seconds
+	t := timerpool.Get(60 * time.Second)
+	select {
+	case c.writeBufferChan <- buf:
+	case <-t.C:
+		buf.Release()
+		err = ErrWriteBufferChanTimeout
+	}
+	timerpool.Put(t)
+	return
+}
+
+func (c *Connection) startReadLoop() {
+	for {
+		select {
+		case <-c.connStopChan:
+			return
+		case <-c.listenerStopChan:
+			return
+		default:
+			n, err := c.doReadIO()
+			if n > 0 {
+				c.onRead(iobufferpool.NewStreamBuffer(c.readBuffer[:n]))
+			}
+
+			if err != nil {
+				if atomic.LoadUint32(&c.closed) == 1 {
+					logger.Debugf("tcp connection has closed, exit read loop, local addr: %s, remote addr: %s",
+						c.localAddr.String(), c.remoteAddr.String())
+					tcpBufferPool.Put(c.readBuffer)
+					return
+				}
+
+				if te, ok := err.(net.Error); ok && te.Timeout() {
+					if n == 0 {
+						continue // continue read data, ignore timeout error
+					}
+				}
+			}
+
+			if err != nil {
+				if err == io.EOF {
+					logger.Debugf("tcp connection remote close, local addr: %s, remote addr: %s, err: %s",
+						c.localAddr.String(), c.remoteAddr.String(), err.Error())
+					_ = c.Close(NoFlush, RemoteClose)
+				} else {
+					logger.Errorf("tcp connection read error, local addr: %s, remote addr: %s, err: %s",
+						c.localAddr.String(), c.remoteAddr.String(), err.Error())
+					_ = c.Close(NoFlush, OnReadErrClose)
+				}
+				return
+			}
+		}
+	}
+}
+
+func (c *Connection) startWriteLoop() {
+	var err error
+	for {
+		select {
+		case <-c.connStopChan:
+			logger.Debugf("connection exit write loop, local addr: %s, remote addr: %s",
+				c.localAddr.String(), c.remoteAddr.String())
+			return
+		case buf, ok := <-c.writeBufferChan:
+			if !ok {
+				return
+			}
+			c.appendBuffer(buf)
+		OUTER:
+			// Keep reading until write buffer channel is full(write buffer channel size is writeBufSize)
+			for i := 0; i < writeBufSize-1; i++ {
+				select {
+				case buf, ok := <-c.writeBufferChan:
+					if !ok {
+						return
+					}
+					c.appendBuffer(buf)
+				default:
+					break OUTER
+				}
+			}
+			_, err = c.doWrite()
+		}
+
+		if err != nil {
```

Comment:
early continue to reduce nesting.

```suggestion
		if err == nil {
		    continue
		 }
```